---
title: "RNAseq"
author: "LIPINSKI Boris"
date: '2025-02'
output:
  rmdformats::readthedown:
    code_folding: hide
    fig.width: 8
    fig.height: 12
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(jsonlite)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(DESeq2)
library(tximportData)
library(tximport)
library(readr)
library(pheatmap)
library(RColorBrewer)
source(file = "/home/bobo/Bureau/Git/script/RNAseq/bin/fonction.r")
script.dir="/home/bobo/Bureau/Git/script/RNAseq/"
```

# Quality check from Kallisto output {.tabset}

This information is to be coupled with the multiQC report generated by the pipeline. The percentage of pseudoaligned reads is more homogeneous and higher when the paired-end protocol is used. Looking at all the figures generated, the results seem to be of better quality when using the paired-end protocol.


```{r, echo=F, warning=F, error=F, message=F}
logs.path <- "/home/bobo/Bureau/Git/result/ALL/QUANTIFICATION/"
logs.files <- list.files(logs.path, pattern = "_*E_run_info.json$", full.names = TRUE)
logs.result <- data.frame()

cols <- c("n_targets", "n_processed", "n_pseudoaligned", "n_unique", "p_pseudoaligned", "p_unique")

for (log in logs.files) {
  log.json <- t(as.data.frame(jsonlite::fromJSON(log)))
  log.json <- as.data.frame(log.json[cols, ])
  colnames(log.json) <- gsub("_run_info.json", "", basename(log))
  if (nrow(logs.result) == 0) {
    logs.result <- log.json
  } else {
    logs.result <- cbind(logs.result, log.json)
  }
}

logs.result <- as.data.frame(t(logs.result))
logs.result$Sample <- rownames(logs.result)
logs.result$Condition <- "Traited"
logs.result[grep("Untreated", logs.result$Sample, ignore.case = TRUE), "Condition" ] <- "Untreated"
logs.result$Protocol <- "PE"
logs.result[grep("SE", logs.result$Sample, ignore.case = TRUE), "Protocol" ] <- "SE"

logs.pivot <- logs.result %>% pivot_longer(cols = -c(Sample,Condition,Protocol), names_to = "Feature", values_to = "Value")
logs.pivot$Value <- as.numeric(logs.pivot$Value)

```

## Protocol {.tabset}
```{r, results='asis', echo=F, warning=F, error=F, message=F}
for (col in cols) {
  cat("### ", col, "\n\n")

  logs.subset <- logs.pivot %>% filter(Feature == col)
  print(ggplot(logs.subset) +
          aes(x = Protocol, y = Value) + 
          geom_boxplot() +
          scale_fill_hue(direction = 1) +
          theme_minimal() +
          facet_wrap(vars(Feature), scales = "free"))
  
  cat("\n\n")
}
```

## Protocol + Condition {.tabset}
```{r, results='asis', echo=F, warning=F, error=F, message=F}
for (col in cols) {
  cat("### ", col, "\n\n")

  logs.subset <- logs.pivot %>% filter(Feature == col)
  moyennes <- logs.subset %>% group_by(Protocol) %>% summarise(moyenne = mean(Value, na.rm = TRUE))
  print(ggplot(logs.subset) + aes(x = Protocol, y = Value, fill = Condition) +
    geom_boxplot() +
    #geom_hline(data = moyennes, aes(yintercept = moyenne, color = Protocol), linetype = "dashed", size = 1) +
    scale_fill_hue(direction = 1) +
    theme_minimal() +
    facet_wrap(vars(Feature), scales = "free") +
    labs(color = "Moyenne par protocole"))

  cat("\n\n")
}
```





# Differental Expression : Treated vs Untreated
```{r, echo=F, warning=F, error=F, message=F}
counts.path <- "/home/bobo/Bureau/Git/result/ALL/QUANTIFICATION/"
counts.files <- list.files(counts.path, pattern = "_*E_abundance.tsv", full.names = TRUE)
counts.result <- data.frame()
#names(counts.files) <- c(
#  "Treated1_PE", "Treated1_SE", "Treated2_PE", "Treated2_SE", "Treated3_PE", "Treated3_SE", "Treated4_PE", "Treated4_SE", 
#  "Untreated1_PE", "Untreated1_SE", "Untreated2_PE", "Untreated2_SE", "Untreated3_PE", "Untreated3_SE", "Untreated4_PE", "Untreated4_SE"
#)
```

```{r, echo=F, warning=F, error=F, message=F}
tx2gene <- read_csv("/home/bobo/Bureau/Git/script/RNAseq/dependencies/tx2gene.gencode.v27.csv")
count.txi <- tximport(counts.files, type = "kallisto", tx2gene = tx2gene, ignoreAfterBar = TRUE)

#metadata <- data.frame(
#  sample = c(
#    "Treated1_PE", "Treated1_SE", "Treated2_PE", "Treated2_SE", "Treated3_PE", "Treated3_SE", "Treated4_PE", "Treated4_SE",
#    "Untreated1_PE", "Untreated1_SE", "Untreated2_PE", "Untreated2_SE", "Untreated3_PE", "Untreated3_SE", "Untreated4_PE", "Untreated4_SE"
#  ),
#  condition = factor(c(rep("Treated",8),rep("Untreated",8) )),
#  protocol = factor(rep(c("PE", "SE"), 8))
#)
#rownames(metadata) <- colnames(count.txi$counts)

metadata <- read_csv(file = "/home/bobo/Bureau/Git/script/RNAseq/dependencies/metadata.csv")
names(counts.files) <- metadata$sample

colnames(count.txi$counts) <- metadata$sample
write_csv(as.data.frame(count.txi$counts), "/home/bobo/Bureau/Git/script/RNAseq/result/all_count.csv")
```




## Single-end {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
dds <- DESeq_process(metadata=metadata, counts.files=counts.files, Protocol_lvl="SE", formule="condition")
res <- DESeq_result(dds)
```

Results :

- DEGs number : `r nrow(res[res$padj < 0.05,])` (padj < 0.05)
- Up-regulated in Treated condition : `r nrow(res[res$Expression =='Up',])`
- Down-regulated in Treated condition : `r nrow(res[res$Expression =='Down',])`
- Positive logFC = up-regulated in Treated condition / down-regulated in Untreated condition
- Negative logFC = up-regulated in Untreated condition / down-regulated in Treated condition

### Dataset {.tabset}
#### Liste {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
dds <- DESeq_process(metadata=metadata, counts.files=counts.files, Protocol_lvl="SE", formule="condition")
res <- DESeq_result(dds)
DT.table(as.data.frame(res[res$padj < 0.05,c("baseMean","log2FoldChange","pvalue","padj","Expression")]), filename="matrice_count")
resSE <- res[res$padj < 0.05,]
```

#### Outlier {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Oulier_norm(dds)
```

#### Heatmap {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
SampleHeatmap(dds)
```

#### PCA {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
PCA_plot(dds)
```


### Expression {.tabset}
#### Volcano {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Volcano_plot(res)
```

#### Top20 {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Top20(res, dds)
```

#### TopHeatmap {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
TopHeatmap(dds, res, FDR=0.01, FC=1.5)
```

#### MDplot {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
MD_plot_DESeq2(dds)
```









## Paired-end {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
dds <- DESeq_process(metadata=metadata, counts.files=counts.files, Protocol_lvl="PE", formule="condition")
res <- DESeq_result(dds)
```

Results :

- DEGs number : `r nrow(res[res$padj < 0.05,])` (padj < 0.05)
- Up-regulated in Treated condition : `r nrow(res[res$Expression =='Up',])`
- Down-regulated in Treated condition : `r nrow(res[res$Expression =='Down',])`
- Positive logFC = up-regulated in Treated condition / down-regulated in Untreated condition
- Negative logFC = up-regulated in Untreated condition / down-regulated in Treated condition



### Dataset {.tabset}
#### Liste {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
dds <- DESeq_process(metadata=metadata, counts.files=counts.files, Protocol_lvl="PE", formule="condition")
res <- DESeq_result(dds)
DT.table(as.data.frame(res[res$padj < 0.05,c("baseMean","log2FoldChange","pvalue","padj","Expression")]), filename="matrice_count")
resPE <- res[res$padj < 0.05,]
```

#### Outlier {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Oulier_norm(dds)
```

#### Heatmap {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
SampleHeatmap(dds)
```

#### PCA {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
PCA_plot(dds)
```


### Expression {.tabset}
#### Volcano {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Volcano_plot(res)
```

#### Top20 {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Top20(res, dds)
```

#### TopHeatmap {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
TopHeatmap(dds, res, FDR=0.01, FC=1.5)
```

#### MDplot {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
MD_plot_DESeq2(dds)
```







# Comparaison Single-end vs Paired-end 

## Intersection of the DEA result

```{r, echo=F, warning=F, error=F, message=F}
library(ggVennDiagram)
x <- list(PE = rownames(resPE), SE = rownames(resSE))
ggVennDiagram(x, category.names = c("Paired-end","Single-end"), label_alpha = 1, label = "count") +
  scale_fill_gradient(low="white",high = "white") + 
  theme(legend.position = "none") + ggtitle("DEA list intersection") + 
  theme(plot.title = element_text(hjust = 0.5)) + coord_flip()
```

There are more differentially expressed transcripts detected between the treated and untreated conditions using the paired-end protocol.
The paired-end protocol allows us to obtain both homogeneous samples and more biological information, as shown in the Kallisto output.

Furthermore, when a PCA is performed with the paired-end protocol, 97% of the dataset's variability is explained by the treated/untreated feature, compared to 89% for the single-end protocol.
This indicates that the biological variability of primary interest (the treatment effect) is much more highlighted and manageable with the paired-end protocol.

Due to the clarity and uniformity of the variability explained by the treatment condition in the paired-end protocol, the DEA results become more meaningful and easier to exploit.
This may also explain why a greater number of differentially expressed transcripts are identified.

The paired-end protocol provides more biological information, which in turn offers greater statistical power, higher confidence in the results, and a clearer signal of the treatment effect compared to the single-end protocol.

Conversely, the single-end protocol introduces more noise into the results, making it less suitable for this type of analysis.

In conclusion, there is a clear treatment effect on the sequenced individuals, regardless of the protocol used. However, this effect is better highlighted when the paired-end protocol is employed.


## Multivariate model (EXPERIMENTAL)
### Dataset {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
dds <- DESeqDataSetFromTximport(count.txi, colData = metadata, design = ~condition)
smallestGroupSize <- 3
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep,]
dds$condition <- factor(dds$condition, levels = c("Untreated","Treated"))
dds <- DESeq(dds)
ddsMF <- dds

# Convertir les colonnes en facteurs
ddsMF$protocol <- as.factor(ddsMF$protocol)
ddsMF$condition <- as.factor(ddsMF$condition)

design(ddsMF) <- formula(~ protocol + condition)
ddsMF <- DESeq(ddsMF)
```

#### Outlier {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Oulier_norm(dds)
```

#### Heatmap {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
SampleHeatmap(dds)
```

#### PCA {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
PCA_plot(dds)
```

#### MDplot {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
MD_plot_DESeq2(dds)
```






### Treated vs Untreated {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
res <- results(ddsMF, contrast=c("condition","Treated","Untreated"))
res <- res[!is.na(res$padj),]
res$threshold <- res$padj < 0.05 ; 
res[res$log2FoldChange > 0 & res$padj < 0.05, "Expression"] <- "Up" 
res[res$log2FoldChange < 0 & res$padj < 0.05, "Expression"] <- "Down"
res[res$padj >= 0.05,  "Expression"] <- "Not Sig"
res <- res[order(res$padj, decreasing = F),]
```

Results :

- DEGs number : `r nrow(res[res$padj < 0.05,])` (padj < 0.05)
- Up-regulated in Treated condition : `r nrow(res[res$Expression =='Up',])`
- Down-regulated in Treated condition : `r nrow(res[res$Expression =='Down',])`
- Positive logFC = up-regulated in Treated condition / down-regulated in Untreated condition
- Negative logFC = up-regulated in Untreated condition / down-regulated in Treated condition




#### Expression {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
DT.table(as.data.frame(res[res$padj < 0.05,c("baseMean","log2FoldChange","pvalue","padj","Expression")]), filename="matrice_count")
```

#### Volcano {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Volcano_plot(res)
```

#### Top20 {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Top20(res, dds)
```

#### TopHeatmap {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
TopHeatmap(dds, res, FDR=0.01, FC=1.5)
```






### Single-end vs Paired-end {.tabset}

```{r, echo=F, warning=F, error=F, message=F}
res <- results(ddsMF, contrast=c("protocol","SE","PE"))
res <- res[!is.na(res$padj),]
res$threshold <- res$padj < 0.05 ; 
res[res$log2FoldChange > 0 & res$padj < 0.05, "Expression"] <- "Up" 
res[res$log2FoldChange < 0 & res$padj < 0.05, "Expression"] <- "Down"
res[res$padj >= 0.05,  "Expression"] <- "Not Sig"
res <- res[order(res$padj, decreasing = F),]
```

Results :

- DEGs number : `r nrow(res[res$padj < 0.05,])` (padj < 0.05)
- Up-regulated in Treated condition : `r nrow(res[res$Expression =='Up',])`
- Down-regulated in Treated condition : `r nrow(res[res$Expression =='Down',])`
- Positive logFC = up-regulated in Treated condition / down-regulated in Untreated condition
- Negative logFC = up-regulated in Untreated condition / down-regulated in Treated condition



#### Expression {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
DT.table(as.data.frame(res[res$padj < 0.05,c("baseMean","log2FoldChange","pvalue","padj","Expression")]), filename="matrice_count")
```

#### Volcano {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Volcano_plot(res)
```

#### Top20 {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
Top20(res, dds)
```

#### TopHeatmap {.tabset}
```{r, echo=F, warning=F, error=F, message=F}
TopHeatmap(dds, res, FDR=0.01, FC=1.5)
```

